src/Sentry/Internal/Http/CachingTransport.cs

```diff
@@ -0,0 +1,266 @@
+                using (var envelope =
+                    await Envelope.DeserializeAsync(envelopeFile, cancellationToken).ConfigureAwait(false))
+                {
+                    _options.DiagnosticLogger?.LogDebug(
+                        "Sending cached envelope: {0}",
+                        envelope.TryGetEventId()
+                    );
+
+                    try
+                    {
+                        await _innerTransport.SendEnvelopeAsync(envelope, cancellationToken).ConfigureAwait(false);
+
+                        _options.DiagnosticLogger?.LogDebug(
+                            "Successfully sent cached envelope: {0}",
+                            envelope.TryGetEventId()
+                        );
+                    }
+                    catch (IOException ex)
+                    {
+                        _options.DiagnosticLogger?.LogError(
+                            "Transient failure when sending cached envelope: {0}",
+                            ex,
+                            envelope.TryGetEventId()
+                        );
+
+                        shouldRetry = true;
+                    }
+                    catch (HttpRequestException ex)
+                    {
+                        _options.DiagnosticLogger?.LogError(
+                            "Persistent failure when sending cached envelope: {0}",
+                            ex,
+                            envelope.TryGetEventId()
+                        );
+                    }
+                }
+
+                // Envelope & file stream must be disposed prior to reaching this point
+
+                if (shouldRetry)
+                {
+                    // Move the file back to cache directory so it gets processed again in the future
+                    Directory.CreateDirectory(_cacheDirectoryPath);
+
+                    File.Move(
+                        envelopeFilePath,
+                        Path.Combine(_cacheDirectoryPath, Path.GetFileName(envelopeFilePath))
+                    );
+
+                    // Avoid processing other files because the order may end up corrupted
+                    break;
+                }
+                else
+                {
+                    // Delete the envelope file and move on to the next one
+                    File.Delete(envelopeFilePath);
+                }
+            }
+        }
+
+        private async ValueTask StoreToCacheAsync(
+            Envelope envelope,
+            CancellationToken cancellationToken = default)
+        {
+            using var lockClaim = await _cacheDirectoryLock.ClaimAsync(cancellationToken).ConfigureAwait(false);
+
+            // If over capacity - remove oldest envelope file
+            while (
+                GetCacheFilePaths().Count() >= _options.MaxQueueItems &&
+                GetCacheFilePaths().FirstOrDefault() is { } oldestEnvelopeFilePath)
```

Comment:
What about assigning the result of `GetCacheFilePaths` before the `FirstOrDefault`?
Maybe a local function would help